### PR TITLE
fix: CasbinJsGetPermissionForUser incompatible with casbin-js

### PR DESCRIPTION
**Pull request by closetool**
Signed-off-by: closetool <c299999999@qq.com>

Fix: #747

### CODE REVIEW

**Comment by closetool** [comment]
@hsluoyz @rickardgranberg @Zxilly plz review

File: frontend.go [hunk 1]
```diff
@@ -15,18 +15,40 @@
...
+	"fmt"
+	"strings"
 )
 
 func CasbinJsGetPermissionForUser(e IEnforcer, user string) ([]byte, error) {
-	policy, err := e.GetImplicitPermissionsForUser(user)
-	if err != nil {
-		return nil, err
+	model := e.GetModel()
+	s := strings.Builder{}
```
Comment (hsluoyz) [inline frontend.go:26]: @closetool plz define `Model.ToText()` method in the model class to get the CONF content from a model struct, return value is string.

Then use it here.

**Review by hsluoyz — CHANGES_REQUESTED** [review]

File: model/model.go [hunk 2]
```diff
@@ -227,3 +227,22 @@ func (model Model) SortPoliciesByPriority() error {
...
+	s.WriteString(fmt.Sprintf("p = %s\n", strings.Replace(model["p"]["p"].Value, "_", ".", -1)))
+	if sec, ok := model["g"]; ok {
+		if ptype, ok := sec["g"]; ok {
+			s.WriteString("[role_definition]\n")
+			s.WriteString(fmt.Sprintf("g = %s\n", ptype.Value))
+		}
+	}
+	s.WriteString("[policy_effect]\n")
+	s.WriteString(fmt.Sprintf("p = %s\n", strings.Replace(model["e"]["e"].Value, "_", ".", -1)))
+	s.WriteString("[matcher]\n")
```
Comment (hsluoyz) [inline model/model.go:245]: matchers

File: model/model.go [hunk 3]
```diff
@@ -227,3 +227,22 @@ func (model Model) SortPoliciesByPriority() error {
...
+	s.WriteString("[policy_definition]\n")
+	s.WriteString(fmt.Sprintf("p = %s\n", strings.Replace(model["p"]["p"].Value, "_", ".", -1)))
+	if sec, ok := model["g"]; ok {
+		if ptype, ok := sec["g"]; ok {
+			s.WriteString("[role_definition]\n")
+			s.WriteString(fmt.Sprintf("g = %s\n", ptype.Value))
+		}
+	}
+	s.WriteString("[policy_effect]\n")
+	s.WriteString(fmt.Sprintf("p = %s\n", strings.Replace(model["e"]["e"].Value, "_", ".", -1)))
```
Comment (hsluoyz) [inline model/model.go:244]: e =

File: model/model.go [hunk 4]
```diff
@@ -227,3 +227,22 @@ func (model Model) SortPoliciesByPriority() error {
...
+	s.WriteString("[matcher]\n")
+	s.WriteString(fmt.Sprintf("p = %s\n", strings.Replace(model["m"]["m"].Value, "_", ".", -1)))
```
Comment (hsluoyz) [inline model/model.go:246]: m =

File: model/model.go [hunk 5]
```diff
@@ -227,3 +227,22 @@ func (model Model) SortPoliciesByPriority() error {
 	}
 	return nil
 }
+
+func (model Model) ToText() string {
+	s := strings.Builder{}
+	s.WriteString("[request_definition]\n")
+	s.WriteString(fmt.Sprintf("r = %s\n", strings.Replace(model["r"]["r"].Value, "_", ".", -1)))
```
Comment (hsluoyz) [inline model/model.go:234]: Plz define `strings.Replace(string, "_", ".", -1)` in a seperated utility function

**Review by hsluoyz — CHANGES_REQUESTED** [review]

**Comment by hsluoyz** [comment]
@Zxilly plz review.

File: model/model.go [hunk 6]
```diff
@@ -227,3 +227,25 @@ func (model Model) SortPoliciesByPriority() error {
...
+	s := strings.Builder{}
+	writeString := func(ptype string) {
+		s.WriteString(fmt.Sprintf("%s = %s\n", ptype, strings.Replace(model[ptype][ptype].Value, "_", ".", -1)))
+	}
+	s.WriteString("[request_definition]\n")
+	writeString("r")
+	s.WriteString("[policy_definition]\n")
+	writeString("p")
+	if sec, ok := model["g"]; ok {
+		if ptype, ok := sec["g"]; ok {
```
Comment (hsluoyz) [inline model/model.go:241]: What about g2, g3?

File: frontend.go [hunk 7]
```diff
@@ -15,18 +15,22 @@
...
 
 func CasbinJsGetPermissionForUser(e IEnforcer, user string) ([]byte, error) {
-	policy, err := e.GetImplicitPermissionsForUser(user)
-	if err != nil {
-		return nil, err
+	model := e.GetModel()
+	m := map[string]interface{}{}
+	m["m"] = model.ToText()
+	policies := e.GetPolicy()
+	for i := range policies {
```
Comment (hsluoyz) [inline frontend.go:27]: Where are g policies?

**Review by hsluoyz — CHANGES_REQUESTED** [review]

**Comment by closetool** [comment]
@hsluoyz I just followed node-casbin's implementation
https://github.com/casbin/node-casbin/blob/master/src/frontend.ts

File: frontend.go (hunk 7 above)
Comment (closetool) [reply frontend.go:27]: @hsluoyz I just followed node-casbin's implementation
https://github.com/casbin/node-casbin/blob/master/src/frontend.ts

Comment (closetool) [reply frontend.go:27]: I guess we should be consistent with node-casbin

Comment (hsluoyz) [reply frontend.go:27]: @Zxilly

Comment (Zxilly) [reply frontend.go:27]: @closetool Yes, seems node-casbin has bug too, it cannot handle multipy role-manager.
In fact, this function means to get the whole model and all policy. You can implement it in golang style. I will later fix node-casbin version.

**Comment by closetool** [comment]
@hsluoyz @Zxilly plz review,
now output likes this:
```json
{"g":[["g2","/book/*","book_group"],["g2","/book/:id","book_group"],["g2","/pen/:id","pen_group"],["g2","/book2/{id}","book_group"],["g2","/pen2/{id}","pen_group"]],"m":"[request_definition]\nr = sub, obj, act\n[policy_definition]\np = sub, obj, act\n[role_definition]\ng2 = _, _\ng = _, _\n[policy_effect]\ne = some(where (p.eft == allow))\n[matchers]\nm = g(r.sub, p.sub) && g2(r.obj, p.obj) && regexMatch(r.act, p.act)\n","p":[["p","alice","/pen/1","GET"],["p","alice","/pen2/1","GET"],["p","book_admin","book_group","GET"],["p","pen_admin","pen_group","GET"]]}
```

**Comment by Zxilly** [comment]
@closetool It expect to return a json string like
`{"m":"[request_definition]\nr = sub, obj, act\n[policy_definition]\np = sub, obj, act\n[role_definition]\ng = _, _\n[policy_effect]\ne = some(where (p.eft == allow))\n[matchers]\nm = g(r.sub, p.sub) && r.obj == p.obj && r.act == p.act","p":[["p","alice","data1","read"],["p","bob","data2","write"],["p","data1_admin","data1","read"],["p","data1_admin","data1","write"],["p","data2_admin","data2","read"],["p","data2_admin","data2","write"]]}`
It will be ok if you use json format.

**Comment by hsluoyz** [comment]
@Zxilly can you run the code to check whether it returns *exactly* the same response as Node-Casbin?

**Comment by Zxilly** [comment]
@closetool I recheck the src of casbin.js. 
```js
public async initEnforcer(s: string): Promise<void> {
        const obj = JSON.parse(s);
        if (!('m' in obj)) {
            throw Error("No model when init enforcer.");
        }
        const m = casbin.newModelFromString(obj['m']);
        this.enforcer = await casbin.newEnforcer(m);
        if ('p' in obj) {
            for (const sArray of obj['p']) {
                await this.enforcer.addPolicy(sArray[1].trim(), sArray[2].trim(), sArray[3].trim());
            }
        }
```
So, there is no need for g section. You only need to send the whole model and parsed policies, that's ok.
Sorry for my vague description.

**Comment by hsluoyz** [comment]
@Zxilly why are `g` policies not needed?

**Comment by Zxilly** [comment]
@hsluoyz it generate roles from model and policies. Just like backend do.

**Comment by closetool** [comment]
@Zxilly Why we do not just get user's permission from backend, instead generate rules by frontend?

**Comment by closetool** [comment]
@Zxilly Why don't we just get user's permission from backend?

**Comment by Zxilly** [comment]
@closetool This is the original design of casbin.js.

**Comment by closetool** [comment]
@Zxilly Is there any thing wrong?
```json
{
    "m":"[request_definition]
r = sub, obj, act
[policy_definition]
p = sub, obj, act
[role_definition]
g = _, _
g2 = _, _
[policy_effect]
e = some(where (p.eft == allow))
[matchers]
m = g(r.sub, p.sub) && g2(r.obj, p.obj) && regexMatch(r.act, p.act)
",
    "p":[
        [
            "p",
            "alice",
            "/pen/1",
            "GET"
        ],
        [
            "p",
            "alice",
            "/pen2/1",
            "GET"
        ],
        [
            "p",
            "book_admin",
            "book_group",
            "GET"
        ],
        [
            "p",
            "pen_admin",
            "pen_group",
            "GET"
        ]
    ]
}
```

**Review by Zxilly — APPROVED** [review]
LGTM

**Review by hsluoyz — APPROVED** [review]

**Pull request merged**